src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/BasicSpringDataRepositoryDatabaseService.java

```diff
@@ -37,105 +42,189 @@
+            if (element.getVersion() >= maxVersion) {
+                maxVersion = element.getVersion();
+                lastVersionEntity = element;
+            }
+        }
+
+        return lastVersionEntity;
     }
 
     @Override
-    public void enableAccessions(List<AccessionWrapper<MODEL, String, ACCESSION>> accessionedObjects) {
-        customMethodsRepository.enableByHashedMessageIn(accessionedObjects.stream().map(AccessionWrapper::getHash)
-                .collect(Collectors.toSet()));
+    public ModelWrapper<MODEL, String, ACCESSION> findAccessionVersion(ACCESSION accession, int version)
+            throws AccessionDoesNotExistException, AccessionDeprecatedException, AccessionMergedException {
+        ACCESSION_ENTITY result = doFindAccessionVersion(accession, version);
+        return toModelWrapper(result);
+    }
+
+    private ACCESSION_ENTITY doFindAccessionVersion(ACCESSION accession, int version) throws
+            AccessionDoesNotExistException, AccessionMergedException, AccessionDeprecatedException {
+        ACCESSION_ENTITY result = repository.findByAccessionAndVersion(accession, version);
+        if (result == null) {
+            assertAccessionMergedOrDeprecated(accession);
+            throw new AccessionDoesNotExistException(accession.toString(), version);
+        }
+        return result;
     }
 
     @Override
-    public AccessionWrapper<MODEL, String, ACCESSION> update(AccessionWrapper<MODEL, String, ACCESSION> accession)
-            throws AccessionDoesNotExistException, HashAlreadyExistsException {
-        Collection<ENTITY> accessionedElements = repository.findByAccession(accession.getAccession());
-        assertAccessionExists(accession, accessionedElements);
-        assertHashDoesNotExist(accession);
+    public void insert(List<ModelWrapper<MODEL, String, ACCESSION>> objects) {
+        Set<ACCESSION_ENTITY> entitySet = objects.stream().map(toEntityFunction).collect(Collectors.toSet());
+        repository.insert(entitySet);
+    }
 
-        int version = 1;
-        for (ENTITY accessionedElement : accessionedElements) {
-            if (version <= accessionedElement.getVersion()) {
-                version = accessionedElement.getVersion() + 1;
+    @Override
+    public AccessionWrapper<MODEL, String, ACCESSION> patch(ModelWrapper<MODEL, String, ACCESSION> accession)
+            throws AccessionDoesNotExistException, HashAlreadyExistsException, AccessionDeprecatedException,
+            AccessionMergedException {
+        List<ACCESSION_ENTITY> entities = getAccession(accession.getAccession());
+        int maxVersion = 1;
+        for (ACCESSION_ENTITY entity : entities) {
+            if (entity.getVersion() >= maxVersion) {
+                maxVersion = entity.getVersion();
             }
         }
+        accession.setVersion(maxVersion + 1);
+        repository.insert(Arrays.asList(toEntityFunction.apply(accession)));
```

Comment:
assertHashDoesNotExist(accession); is missing

Reply:
Good point, I wanted to make it fail before proceed on the other case and in this case I've left it so that the database is the one that handles that. I'm gonna do a double check and try to unify criteria.

Reply:
I think that should be covered by the `insert` method, which will fail due to a constraint violation. But it is true that for error reporting purposes it could be nice to always use the same exceptions.